src/backend/distributed/commands/schema_based_sharding.c

```diff
@@ -108,6 +122,94 @@ ShouldCreateTenantSchemaTable(Oid relationId)
 }
 
 
+/*
+ * EnsureTableKindSupportedForTenantSchema ensures that given table's kind is
+ * supported by a tenant schema.
+ */
+static void
+EnsureTableKindSupportedForTenantSchema(Oid relationId)
+{
+	if (IsForeignTable(relationId))
+	{
+		ereport(ERROR, (errmsg("cannot create a foreign table in a distributed "
+							   "schema")));
+	}
+	else if (PartitionTable(relationId))
+	{
+		ErrorIfIllegalPartitioningInTenantSchema(PartitionParentOid(relationId),
+												 relationId);
+	}
+	else if (IsChildTable(relationId) || IsParentTable(relationId))
+	{
+		ereport(ERROR, (errmsg("tables in a distributed schema cannot inherit or "
+							   "be inherited")));
+	}
+}
+
+
+/*
+ * EnsureFKeysForTenantTable ensures that all referencing and referenced foreign
+ * keys are allowed for given table.
+ */
+static void
+EnsureFKeysForTenantTable(Oid relationId)
+{
+	Oid tenantSchemaId = get_rel_namespace(relationId);
+	int fKeyReferencingFlags = INCLUDE_REFERENCING_CONSTRAINTS | INCLUDE_ALL_TABLE_TYPES;
+	List *referencingForeignKeys = GetForeignKeyOids(relationId, fKeyReferencingFlags);
+	Oid foreignKeyId = InvalidOid;
+	foreach_oid(foreignKeyId, referencingForeignKeys)
+	{
+		Oid referencedTableId = GetReferencedTableId(foreignKeyId);
+		Oid referencedTableSchemaId = get_rel_namespace(referencedTableId);
+
+		/* We allow foreign keys to a table in the same schema */
+		if (tenantSchemaId == referencedTableSchemaId)
+		{
+			continue;
+		}
+
+		/*
+		 * Allow foreign keys to the other schema only if the referenced table is
+		 * a reference table.
+		 */
+		if (!IsCitusTable(referencedTableId) ||
+			!IsCitusTableType(referencedTableId, REFERENCE_TABLE))
+		{
+			ereport(ERROR, errmsg("when referenced table is in another schema, "
+								  "only foreign keys to a reference table is "
+								  "allowed from a table in distributed schema"));
```

Comment:
A suggestion to make the error message a bit more clear:

```suggestion
			ereport(ERROR, errmsg("foreign keys from tables in distributed "
			                      "schemas are not allowed when the referenced "
			                      "table is in another schema and not a reference "
			                      "table"),
			               errdetail("\"%s\" references \"%s\" via foreign key "
			                         "constraint" \"%s\",
			                         generate_qualified_relation_name, 
			                         generate_qualified_relation_name,
			                         get_constraint_name));
```

Does that look good @aykut-bozkurt /  @mulander?